File: Лабораторная 2/WindowsFormsApp5/Liner.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WindowsFormsApp5
{
    public class Liner : PassengerShip, IComparable<Liner>, IEquatable<Liner>
    {
        private Color dopColor;
        private bool pool;
        private bool palyb;
        private bool aquapark;
        private bool restaurant;

        public int CompareTo(Liner other)
        {
            var res = (this is PassengerShip).CompareTo(other is PassengerShip);
            if (res!=0)
            {
                return res;
            }
            if (pool!=other.pool)
            {
                return pool.CompareTo(other.pool);
            }
            if(palyb!=other.palyb)
            {
                return palyb.CompareTo(other.palyb);
            }
            if (restaurant!=other.restaurant)
            {
                return restaurant.CompareTo(other.restaurant);
            }
            if (aquapark!=other.aquapark)
            {
                return aquapark.CompareTo(other.aquapark);
            }
            if (ColorBody != other.ColorBody)
            {
                return ColorBody.Name.CompareTo(other.ColorBody);
            }
            if (dopColor!=other.dopColor)
            {
                return dopColor.Name.CompareTo(other.dopColor.Name);
            }
            return 0;
        }
        public bool Equals(Liner other)
        {
            var res = (this is PassengerShip).Equals(other is PassengerShip);
            if (!res )
            {
                return res;
            }
            if (pool != other.pool)
            {
                return false;
            }
            if (palyb != other.palyb)
            {
                return false;
            }
            if (restaurant != other.restaurant)
            {
                return false;
            }
            if (aquapark != other.aquapark)
            {
                return false;
            }
            if (ColorBody != other.ColorBody)
            {
                return false;
            }
            if (dopColor != other.dopColor)
            {
                return false;
            }
            return true;
        }
        public override bool Equals(object obj)
        {
            if(obj==null)
            {
                return false;
            }
            Liner shipObj = obj as Liner;
            if (shipObj==null)
            {
                return false;
            }
            else
            {
                return Equals(shipObj);
            }
        }
        public override int GetHashCode()
        {
            return base.GetHashCode();
        }
        public Liner(int maxSpeed, int maxCountPassengers, double weight, Color color, bool pool, bool palyb, bool aquapark, bool restaurant, Color dopColor) : base(maxCountPassengers, maxSpeed, weight, color)
        {
            this.dopColor = dopColor;
            this.pool = pool;
            this.palyb = palyb;
            this.aquapark = aquapark;
            this.restaurant = restaurant;
        }
        public Liner(string info) : base(info)
        {
            string[] strs = info.Split(';');
            if (strs.Length == 9)
            {
                MaxSpeed = Convert.ToInt32(strs[0]);
                MaxCountPassengers = Convert.ToInt32(strs[1]);
                Weight = Convert.ToInt32(strs[2]);
                ColorBody = Color.FromName(strs[3]);
                pool = Convert.ToBoolean(strs[4]);
                palyb = Convert.ToBoolean(strs[5]);
                aquapark = Convert.ToBoolean(strs[6]);
                restaurant = Convert.ToBoolean(strs[7]);
                dopColor = Color.FromName(strs[8]);

            }
        }
        protected override void drawUsualShip(Graphics g)
        {

            if (pool)
            {
                Brush brBlue = new SolidBrush(dopColor);
                g.FillEllipse(brBlue, startPosX + 20, startPosY + 45, 10, 10);
                g.FillEllipse(brBlue, startPosX + 40, startPosY + 45, 10, 10);
                g.FillEllipse(brBlue, startPosX + 60, startPosY + 45, 10, 10);
            }
            base.drawUsualShip(g);

            if (aquapark)
            {
                Pen pen = new Pen(dopColor);
                g.DrawLine(pen, startPosX + 30, startPosY + 40, startPosX + 30, startPosY + 10);
                g.DrawLine(pen, startPosX + 30, startPosY + 10, startPosX + 50, startPosY + 40);
                g.DrawLine(pen, startPosX + 30, startPosY + 10, startPosX + 20, startPosY + 40);

            }

            if (palyb)
            {

                Brush br = new SolidBrush(dopColor);
                g.FillRectangle(br, startPosX + 20, startPosY + 60, 60, 5);
            }
            if (restaurant)
            {

                Brush br = new SolidBrush(dopColor);
                g.FillRectangle(br, startPosX, startPosY + 20, 20, 20);
            }

        }
        public void SetDopColor(Color color)
        {
            dopColor = color;
        }
        public override string getInfo()
        {
            return MaxSpeed + ";" + MaxCountPassengers + ";" + Weight + ";" + ColorBody.Name + ";" + pool + ";" + aquapark + ";" + palyb + ";" + restaurant + ";" + dopColor.Name;

        }
    }
}

File: lab_2_TP/src/lab_2_TP/Liner.java
package lab_2_TP;

import java.awt.Color;
import java.awt.Graphics;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;

public class Liner extends PassengerShip implements Serializable {
	private boolean pool;
	private boolean palyb;
	private boolean aquapark;
	private boolean restaurant;
	transient private Color dopColor;

	public Liner(int maxSpeed, int maxCountPassengers, int weight, Color color,
			boolean pool, boolean palyb, boolean aquapark, boolean restaurant,
			Color dopColor) {
		super(maxCountPassengers, maxSpeed, weight, color);
		this.pool = pool;
		this.palyb = palyb;
		this.aquapark = aquapark;
		this.restaurant = restaurant;
		this.dopColor = dopColor;
	}

	protected void drawUsualShip(Graphics g) {
		super.drawUsualShip(g);

		if (pool) {
			g.setColor(dopColor);
			g.fillOval(startPosX + 20, startPosY + 45, 10, 10);
			g.fillOval(startPosX + 40, startPosY + 45, 10, 10);
			g.fillOval(startPosX + 60, startPosY + 45, 10, 10);
		}

		if (aquapark) {
			g.setColor(dopColor);
			g.drawLine(startPosX + 30, startPosY + 40, startPosX + 30,
					startPosY + 10);
			g.drawLine(startPosX + 30, startPosY + 10, startPosX + 50,
					startPosY + 40);
			g.drawLine(startPosX + 30, startPosY + 10, startPosX + 20,
					startPosY + 40);

		}

		if (palyb) {

			g.setColor(dopColor);
			g.fillRect(startPosX + 20, startPosY + 60, 60, 5);
		}
		if (restaurant) {

			g.setColor(dopColor);
			g.fillRect(startPosX, startPosY + 20, 20, 20);
		}
	}

	public void setDopColor(Color f) {
		dopColor = f;
	}
	private void writeObject(ObjectOutputStream s) throws IOException {
		        s.defaultWriteObject();
		         s.writeInt(ColorBody.getRed());
		         s.writeInt(ColorBody.getGreen());
		         s.writeInt(ColorBody.getBlue());
		         s.writeInt(dopColor.getRed());
		         s.writeInt(dopColor.getGreen());
		         s.writeInt(dopColor.getBlue());
		     }
		 
		     private void readObject(ObjectInputStream s) throws IOException, ClassNotFoundException {
		         s.defaultReadObject();
		         int red = s.readInt();
		         int green = s.readInt();
		         int blue = s.readInt();
		         ColorBody = new Color(red, green, blue);
		         int red1 = s.readInt();
		         int green1 = s.readInt();
		         int blue1 = s.readInt();
		         dopColor = new Color(red1, green1, blue1);
		     }
		 
		     public Liner(String info){
		     	     	super(info);
		     	         String[] strs = info.split(";");
		     	         if (strs.length == 7)
		     	         {
		     	        	 MaxSpeed = Integer.parseInt(strs[0]);
			   	             MaxCountPassengers = Integer.parseInt(strs[1]);
			   	             ColorBody = Color.decode(strs[2]);
			   	             Weight = Integer.parseInt(strs[3]);
		   	             countPassengers = Integer.parseInt(strs[4]);  	            
		     	             restaurant = Boolean.parseBoolean(strs[5]);
		     	             aquapark = Boolean.parseBoolean(strs[6]);
		     	             palyb = Boolean.parseBoolean(strs[7]);
		     	             pool = Boolean.parseBoolean(strs[8]);
		     	             dopColor= Color.decode(strs[9]);
		     	             
		     	         }
		     	     }
		     
		     
		 	public String getInfo() {
		 		// TODO Auto-generated method stub
		 		return MaxSpeed + ";" + MaxCountPassengers + ";" + Weight + ";" + countPassengers + ";" + restaurant + ";" + aquapark + ";" + palyb + ";" +pool+";"+ dopColor;
		 
		 	}
}
